### PR TITLE
Notify

### PR DESCRIPTION
**Pull request by kylealexanderhackney**
Adds sending of emails for
- high cost variations (raising, approval, rejection)
- high cost work order (raising, approval, rejection)
- Adds debuggable log stream into the integration tests

### CODE REVIEW

File: RepairsApi.Tests/E2ETests/Repairs/RepairApiTests.cs [hunk 1]
```diff
@@ -22,6 +22,7 @@
 using WorkOrderComplete = RepairsApi.V2.Generated.WorkOrderComplete;
 using RepairsApi.V2.Services;
 using V2_Generated_DRS;
+using RepairsApi.V2.Email;
```
Comment (kylealexanderhackney) [inline RepairsApi.Tests/E2ETests/Repairs/RepairApiTests.cs:25]: Revert

File: RepairsApi.Tests/E2ETests/Repairs/RepairApiVariationTests.cs [hunk 2]
```diff
@@ -27,22 +27,38 @@ public async Task UpdateCausesPendingApprovalOnWorkOrderWhenCostLimitExceeded()
             // Assert
             workOrder.StatusCode.Should().Be(WorkStatusCode.VariationPendingApproval);
             workOrder.JobStatusUpdates[0].TypeCode.Should().Be(JobStatusUpdateTypeCode._180);
+
+            VerifyEmailSent("highCostVariation");
```
Comment (kylealexanderhackney) [inline RepairsApi.Tests/E2ETests/Repairs/RepairApiVariationTests.cs:31]: Constants for test guids

File: RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/ApproveVariationTests.cs [hunk 3]
```diff
@@ -44,7 +44,7 @@ public void Setup()
 
             _classUnderTest = new ApproveVariationUseCase(
                 _repairsGatewayMock.Object, _jobStatusUpdateGateway.Object,
-                _currentUserServiceMock.Object, _updateSorCodesUseCaseMock.Object);
+                _currentUserServiceMock.Object, _updateSorCodesUseCaseMock.Object, new NotificationMock());
```
Comment (kylealexanderhackney) [inline RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/ApproveVariationTests.cs:47]: test handlers

File: RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/MoreSpecificSorCodeTests.cs [hunk 4]
```diff
@@ -50,7 +50,8 @@ public void Setup()
                 _featureManagerMock.Object,
                 _currentUserServiceMock.Object,
                 _updateSorCodesUseCaseMock.Object,
-                _sheduleOfRatesGateway.Object);
+                _sheduleOfRatesGateway.Object,
```
Comment (kylealexanderhackney) [inline RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/MoreSpecificSorCodeTests.cs:53]: unit tests

File: RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/RejectVariationTests.cs [hunk 5]
```diff
@@ -26,6 +26,7 @@ public class RejectVariationTests
 
         private MockRepairsGateway _repairsGatewayMock;
         private CurrentUserServiceMock _currentUserServiceMock;
+        private Mock<IJobStatusUpdateGateway> _jobStatusUpdateGatewayMock;
```
Comment (kylealexanderhackney) [inline RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/RejectVariationTests.cs:29]: unit test

File: RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/RejectWorkOrderStrategyTests.cs [hunk 6]
```diff
@@ -30,7 +30,8 @@ public void Setup()
             _fixture.Behaviors.Add(new OmitOnRecursionBehavior());
             _currentUserServiceMock = new CurrentUserServiceMock();
             _classUnderTest = new RejectWorkOrderStrategy(
-                _currentUserServiceMock.Object
```
Comment (kylealexanderhackney) [inline RepairsApi.Tests/V2/UseCase/JobStatusUpdateUseCases/RejectWorkOrderStrategyTests.cs:33]: unit tests

File: RepairsApi/V2/Email/GovUKNotifyService.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Notify.Client;
+using Notify.Exceptions;
+using Notify.Interfaces;
+using System.Threading.Tasks;
+
+namespace RepairsApi.V2.Email
+{
+    public interface IEmailService
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Email/GovUKNotifyService.cs:10]: new file

File: RepairsApi/V2/Email/GovUKNotifyService.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            _notificationClientAsync = notificationClientAsync;
+        }
+
+        public Task SendMailAsync<TRequest>(TRequest request)
+            where TRequest : EmailRequest
+        {
+            try
+            {
+                var requestName = typeof(TRequest).Name;
+                if (_options.TemplateIds.TryGetValue(requestName, out var templateId))
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Email/GovUKNotifyService.cs:37]: try cleaning this up

File: RepairsApi/V2/Email/GovUKNotifyService.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+                }
+            }
+            catch (NotifyClientException ex)
+            {
+                _logger.LogError(ex, "[EmailError] Failed to send Email with error {ex.Message}", ex.Message);
+                return Task.CompletedTask;
+            }
+        }
+
+        private string ResolveTemplateId<TRequest>()
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Email/GovUKNotifyService.cs:54]: unused atm

File: RepairsApi/V2/Email/HighCostWorkOrderEmail.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿namespace RepairsApi.V2.Email
+{
+    internal class HighCostWorkOrderEmail : EmailRequest
+    {
+        public HighCostWorkOrderEmail(string address, int id)
+            : base (address)
+        {
+            Set("workOrderId", id);
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Email/HighCostWorkOrderEmail.cs:8]: use constants

File: RepairsApi/V2/Email/VariationEmailNotificationHandler.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            IScheduleOfRatesGateway scheduleOfRatesGateway)
+        {
+            _lazyGovUkService = lazyEmailService;
+            _featureManager = featureMananger;
+            _logger = logger;
+            _scheduleOfRatesGateway = scheduleOfRatesGateway;
+        }
+
+        private async Task SendMail<T>(T emailRequest, Action logAction)
+            where T : EmailRequest
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Email/VariationEmailNotificationHandler.cs:35]: Move to helper class  as its common to both handlers

File: RepairsApi/V2/Email/VariationEmailNotificationHandler.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            if (!await _featureManager.IsEnabledAsync(FeatureFlags.NotifyIntegration))
+            {
+                return;
+            }
+
+            await EmailService.SendMailAsync(emailRequest);
+
+            logAction();
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Email/VariationEmailNotificationHandler.cs:44]: use log scope and log attempt

File: RepairsApi/V2/Gateways/IScheduleOfRatesGateway.cs [hunk 13]
```diff
@@ -20,6 +20,6 @@ public interface IScheduleOfRatesGateway
         Task<IEnumerable<Contractor>> GetLiveContractors();
         Task<Contractor> GetContractor(string contractorReference);
         Task<ScheduleOfRatesModel> GetCode(string sorCode, string propertyReference, string contractorReference);
+        Task<string> GetContractManagerEmail(string contractorReference);
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Gateways/IScheduleOfRatesGateway.cs:23]: Test this

File: RepairsApi/V2/Notifications/HighCostWorkOrderCreated.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace RepairsApi.V2.Notifications
+{
+    public class HighCostWorkOrderCreated : INotification
+    {
+        public HighCostWorkOrderCreated(WorkOrder workOrder)
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/Notifications/HighCostWorkOrderCreated.cs:11]: prop above

File: RepairsApi/V2/UseCase/CreateWorkOrderUseCase.cs [hunk 15]
```diff
@@ -81,6 +81,8 @@ private async Task<WorkOrderOpened> NotifyHandlers(WorkOrder workOrder)
         {
             if (workOrder.StatusCode != WorkStatusCode.Open)
             {
+                await _notifier.Notify(new HighCostWorkOrderCreated(workOrder));
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/UseCase/CreateWorkOrderUseCase.cs:84]: test this

File: RepairsApi/V2/UseCase/JobStatusUpdatesUseCases/ApproveWorkOrderStrategy.cs [hunk 16]
```diff
@@ -41,6 +41,8 @@ private async Task NotifyHandlers(Infrastructure.WorkOrder workOrder)
         {
             var notification = new WorkOrderOpened(workOrder);
             await _notifier.Notify(notification);
+
+            await _notifier.Notify(new WorkOrderApproved(workOrder));
```
Comment (kylealexanderhackney) [inline RepairsApi/V2/UseCase/JobStatusUpdatesUseCases/ApproveWorkOrderStrategy.cs:45]: unit tests

**Review by kylealexanderhackney — COMMENTED** [review]
- Make sure feature works when no config is provided
- test guid config via env parameters
- check code coverage
- Migration
- setup/test env config

File: RepairsApi/V2/Email/GovUKNotifyService.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        public async Task SendMailAsync<TRequest>(TRequest request)
+            where TRequest : EmailRequest
+        {
+            try
+            {
+                var requestName = typeof(TRequest).Name;
+                if (_options.TemplateIds.TryGetValue(requestName, out var templateId))
+                {
+                    await _notificationClientAsync.SendEmailAsync(request.Address, templateId, request);
+                    _logger.LogInformation("Successfully Sent Email {requestName}", requestName);
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/GovUKNotifyService.cs:35]: Resource Strings?

File: RepairsApi/V2/Email/GovUKNotifyService.cs [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+                }
+                else
+                {
+                    _logger.LogError("[EmailError] Could not resolve template for email {requestName}", requestName);
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/GovUKNotifyService.cs:39]: Resource Strings?

File: RepairsApi/V2/Email/GovUKNotifyService.cs [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+                }
+            }
+            catch (NotifyClientException ex)
+            {
+                _logger.LogError(ex, "[EmailError] Failed to send Email with error {ex.Message}", ex.Message);
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/GovUKNotifyService.cs:44]: Resource Strings?

File: RepairsApi/V2/Email/VariationEmailNotificationHandler.cs [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+            {
+                logAction();
+                await EmailService.SendMailAsync(emailRequest);
+            }
+        }
+
+        public Task Notify(VariationRejected data)
+        {
+            return SendMail(new VariationRejectedEmail(data.Variation.AuthorEmail, data.Rejection.RelatedWorkOrder.Id),
+                () => _logger.LogInformation("Sending Mail for rejection of high cost variation on work order {WorkOrderId}", data.Rejection.RelatedWorkOrder.Id));
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/VariationEmailNotificationHandler.cs:52]: Resource Strings?

File: RepairsApi/V2/Email/VariationEmailNotificationHandler.cs [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            var emailAddress = await _scheduleOfRatesGateway.GetContractManagerEmail(data.WorkOrder.AssignedToPrimary.ContractorReference);
+
+            if (string.IsNullOrWhiteSpace(emailAddress))
+            {
+                return;
+            }
+
+            await SendMail(new HighCostVariationCreatedEmail(emailAddress, data.WorkOrder.Id),
+                () => _logger.LogInformation("Sending Mail for raising of high cost variation on work order {WorkOrderId}", data.WorkOrder.Id));
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/VariationEmailNotificationHandler.cs:65]: Resource Strings?

File: RepairsApi/V2/Email/VariationEmailNotificationHandler.cs [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        public Task Notify(VariationApproved data)
+        {
+            return SendMail(new VariationApprovedEmail(data.Variation.AuthorEmail, data.Approval.RelatedWorkOrder.Id),
+                () => _logger.LogInformation("Sending Mail for approval of high cost variation on work order {WorkOrderId}", data.Approval.RelatedWorkOrder.Id));
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/VariationEmailNotificationHandler.cs:71]: Resource Strings?

File: RepairsApi/V2/Email/WorkOrderEmailNotificationHandler.cs [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                logAction();
+                await EmailService.SendMailAsync(emailRequest);
+            }
+        }
+
+        public Task Notify(HighCostWorkOrderCreated data)
+        {
+            return SendMail(new HighCostWorkOrderEmail(_options.Value.PendingWorkOrderRecipient, data.WorkOrder.Id),
+                () => _logger.LogInformation("Sending Mail for creation of high cost work order {WorkOrderId}", data.WorkOrder.Id));
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/WorkOrderEmailNotificationHandler.cs:52]: Resource Strings?

File: RepairsApi/V2/Email/WorkOrderEmailNotificationHandler.cs [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public Task Notify(WorkOrderApproved data)
+        {
+            return SendMail(new WorkApprovedEmail(data.WorkOrder.AgentEmail, data.WorkOrder.Id),
+                () => _logger.LogInformation("Sending Mail for approval of high cost work order {WorkOrderId}", data.WorkOrder.Id));
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/WorkOrderEmailNotificationHandler.cs:58]: Resource Strings?

File: RepairsApi/V2/Email/WorkOrderEmailNotificationHandler.cs [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public Task Notify(WorkOrderRejected data)
+        {
+            return SendMail(new WorkRejectedEmail(data.WorkOrder.AgentEmail, data.WorkOrder.Id),
+                () => _logger.LogInformation("Sending Mail for rejection of high cost work order {WorkOrderId}", data.WorkOrder.Id));
```
Comment (josh-dadswell-hackney) [inline RepairsApi/V2/Email/WorkOrderEmailNotificationHandler.cs:64]: Resource Strings?

**Review by josh-dadswell-hackney — APPROVED** [review]

**Pull request merged**